core/unit_test/incremental/Test14_MDRangeReduce.hpp

```diff
@@ -0,0 +1,180 @@
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+/// @Kokkos_Feature_Level_Required:14
+// Incremental test for MDRange reduction .
+// Reduction is tested with scalar, view and a customized reduction.
+
+#include <Kokkos_Core.hpp>
+#include <gtest/gtest.h>
+
+namespace Test {
+using value_type = double;
+const int N      = 10;
+const int M      = 10;
+
+// A structure for complex number.
+struct MyComplex {
+  value_type _re, _im;
+
+  MyComplex() = default;
+
+  MyComplex(value_type re, value_type im) : _re(re), _im(im) {}
+
+  KOKKOS_INLINE_FUNCTION
+  MyComplex(const MyComplex& src) : _re(src._re), _im(src._im) {}
+
+  KOKKOS_INLINE_FUNCTION
+  void operator+=(const MyComplex& src) {
+    _re += src._re;
+    _im += src._im;
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  void operator+=(const volatile MyComplex& src) volatile {
+    _re += src._re;
+    _im += src._im;
+  }
+};
+
+template <class ExecSpace>
+struct TestMDRangeReduce {
+  // 1D  View of double
+  using View_1D = typename Kokkos::View<value_type*, ExecSpace>;
+
+  // 2D  View of double
+  using View_2D = typename Kokkos::View<value_type**, ExecSpace>;
+
+  // Index Type for the iterator
+  using int_index = Kokkos::IndexType<int>;
+
+  // An MDRangePolicy for 2 nested loops
+  using MDPolicyType_2D = typename Kokkos::Experimental::MDRangePolicy<
+      ExecSpace, Kokkos::Experimental::Rank<2>, int_index>;
+
+  //  1D - complex View
+  using complex_View_1D = typename Kokkos::View<MyComplex*, ExecSpace>;
+
+  // Reduction when ExecPolicy = MDRangePolicy and ReducerArgument =
+  // scalar/1-element view
+  void reduce_MDRange() {
+    View_2D d_data("d_data", N, M);
+
+    MDPolicyType_2D mdPolicy_2D({0, 0}, {N, M});
+
+    // Store the reduced value.
+    value_type d_result = 0.0, h_result = 0.0, view_results = 0.0;
+    Kokkos::View<value_type, ExecSpace> d_resultView("result View");
+
+    // Fill host data with intial values
+    for (int i = 0; i < N; ++i)
+      for (int j = 0; j < M; ++j) h_result += i * j;
+    h_result *= 0.5;
+
+    // Fill data.
+    Kokkos::parallel_for(
+        mdPolicy_2D, KOKKOS_LAMBDA(const int i, const int j) {
+          d_data(i, j) = i * j * 0.5;
+        });
+
+    // Parallel reduce on a scalar.
+    Kokkos::parallel_reduce(
+        mdPolicy_2D,
+        KOKKOS_LAMBDA(const int i, const int j, value_type& update_value) {
+          update_value += d_data(i, j);
+        },
+        d_result);
+
+    // Parallel reduce on a view.
+    Kokkos::parallel_reduce(
+        mdPolicy_2D,
+        KOKKOS_LAMBDA(const int i, const int j, value_type& update_value) {
+          update_value += d_data(i, j);
+        },
+        d_resultView);
+
+    // Check correctness.
+    ASSERT_EQ(h_result, d_result);
+
+    // Copy view back to host.
+    Kokkos::deep_copy(view_results, d_resultView);
```

Comment:
Why is `view_results` plural.  Avoid declaring it so far away from the place you actually use it.